apps/evaluations/utils.py

```diff
@@ -165,3 +168,55 @@ def _clean_context_field_name(field_name):
     field_name = re.sub(r"_+", "_", field_name).strip("_")
 
     return field_name or "context_variable"
+
+
+def make_message_pairs_from_queryset(queryset: QuerySet) -> list[ChatMessage]:
+    """Takes a queryset of ChatMessages, and adds extra messages such that we always have (Human, AI) message pairs.
+    There can be a single AI Message at the beginning (AI seed message).
+    All messages must be from the same chat.
+    """
+
+    if not queryset:
+        return []
+
+    queryset_list = list(queryset)
+    chat = queryset_list[0].chat
+
+    all_chat_messages = list(chat.messages.order_by("created_at"))
+    message_lookup = {msg.id: i for i, msg in enumerate(all_chat_messages)}
+
+    def get_previous_message(message):
+        current_index = message_lookup[message.id]
+        return all_chat_messages[current_index - 1] if current_index > 0 else None
+
+    def get_next_message(message):
+        current_index = message_lookup[message.id]
+        return all_chat_messages[current_index + 1] if current_index < len(all_chat_messages) - 1 else None
+
+    all_messages = set()
+
+    # Handle AI seed message
+    has_ai_seed_message = False
+    if queryset_list[0].is_first_message and queryset_list[0].is_ai_message:
```

Comment:
Could you not annotate the queryset with `is_first_message` and inside the loop (line 205) check if it is the first message. This will allow us to use an iterator for the loop incase the queryset is huge, save us from having to fetch all records on line 182 and also not require an extra DB query to check if the message is the first one.